com.unity.netcode.adapter.utp/Runtime/UnityTransport.cs

```diff
@@ -517,13 +517,67 @@ private void Update()
                 {
                     ;
                 }
+                
+#if MULTIPLAYER_TOOLS
+                if (NetworkManager.Singleton.IsServer)
+                {
+                    var ngoConnectionIds = NetworkManager.Singleton.ConnectedClients.Keys;
+                    foreach (var ngoConnectionId in ngoConnectionIds)
+                    {
+                        ExtractNetworkMetrics(NetworkManager.Singleton.ClientIdToTransportId(ngoConnectionId));
+                    }
+                }
+                else
+                {
+                    ExtractNetworkMetrics(NetworkManager.Singleton.ClientIdToTransportId(NetworkManager.Singleton.ServerClientId));
+                }
+#endif
             }
         }
 
         private void OnDestroy()
         {
             DisposeDriver();
         }
+        
+#if MULTIPLAYER_TOOLS
+        private void ExtractNetworkMetrics(ulong transportClientId)
+        {
+            var networkConnection =  ParseClientId(transportClientId);
+            
+            //Two pipelines, unreliable sequence and reliable fragmented
+            for (byte i = 0; i < 2; ++i)
+            {
+                var pipeline = i == 0 ? m_UnreliableSequencedPipeline : m_ReliableSequencedFragmentedPipeline;
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+                try
+#endif
+                {
+                    //Don't need to dispose of the buffers, they are filled with data pointers. 
+                    m_Driver.GetPipelineBuffers(pipeline, NetworkPipelineStageCollection.GetStageId(typeof(NetworkMetricsPipelineStage)),
+                        networkConnection, out var readProcessingBuffer, out var writeProcessingBuffer, out var sharedBuffer);
+
+                    unsafe
+                    {
+                        var networkMetricsContext = (NetworkMetricsContext*)sharedBuffer.GetUnsafePtr();
+
+                        NetworkMetrics.TrackPacketSent(networkMetricsContext->PacketSentCount);
+                        NetworkMetrics.TrackPacketReceived(networkMetricsContext->PacketReceivedCount);
+
+                        networkMetricsContext->PacketSentCount = 0;
+                        networkMetricsContext->PacketReceivedCount = 0;
+                    }
+                }
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+                catch (InvalidOperationException e)
+                {
+                    //Can ignore, may happen if a pipeline is not used
```

Comment:
Concerned about this - we ignore the exception when ENABLE_UNITY_COLLECTION_CHECKS is enabled, but when it's not enabled it will throw right? What happens when it throws?

Reply:
This gets throwned only if this is defined.

Reply:
Ohh I see. That's worth a comment I think to avoid others getting confused the same (if you keep the try/catch at least, I know simon has called it into question in another comment)